### ISSUE TITLE
Add cargo watch

### ISSUE THREAD

**Issue by Zerotask** [rust-lang/cargo#9339]
**Describe the problem you are trying to solve**

During development you are very often in the process of making changes and then build or run your project or run your tests. This always costs some time, therefore an automation would be useful, to save some time and increase the developer experience

**Describe the solution you'd like**

I'd like to have a built-in watcher similar to nodemon for NodeJS or like this project https://github.com/passcod/cargo-watch
This would allow you to run your whole project if you change something or you can also specify to run only tests when your files change:
`cargo watch -x run`
`cargo watch -x test`

Since this is completely optional to use, developers are not forced to use the suggested watcher.

**Notes**

**Comment by sr2ds** [comment]
@Zerotask, I usage the cargo-watch package and is like nodemon. Why do you think that is necessary include this on cargo if the package cargo-watch solve this problem?

When I need make any tests without cargo, I run on homemade script like this:

```
while inotifywait -e close_write *; do rustc main.rs; done
```

Sorry if I did not understand correctly your request.

**Comment by Zerotask** [comment]
Yes, I also mentioned this package and the reason why I think it should be added to the default toolchain of cargo.

If a command is often used, one could consider to add it to the default toolchain.

**Comment by sr2ds** [comment]
Yes! I understand your wish now and this make sense. I use every day cargo watch too.

**Comment by weihanglo** [comment]
Just note that Cargos has a function [`list_files`] respecting `package.include` and `package.exclude`. This might be worth a reuse than inventing a brand new file listing mechanism. There are also [change detection instructions] in build scripts we need to consider. O

To be fair, I don't know any details of cargo-watch. [epage's suggestion] seems like a great starting point to me.

[`list_files`]: https://github.com/rust-lang/cargo/blob/777960615afe01ca3a822398e645027f2aeb2ce3/src/cargo/sources/path.rs#L100
[change detection instructions]: https://doc.rust-lang.org/cargo/reference/build-scripts.html#change-detection
[epage's suggestion]: https://internals.rust-lang.org/t/cargo-run-watch-flag-to-auto-recompile/17971/4?u=weihanglo

**Comment by epage** [comment]
Another thing I'd add is to study multiple forms of prior art for features, pitfalls, etc so and report on them and what you suggest we should apply from them.

We do this when adding commands but I suspect a `--watch` flag has enough little pieces of policy to be worth it as well.

**Comment by epage** [comment]
While this might only be implemented a bit at at time, we should consider what all commands should support this.

Speaking of a bit at a time, a plan for implementing this, at least from MVP to what you care about.

**Comment by reinaldorauch** [comment]
I think the MVP should only care about of watching for files changes under src (or respectives if not a binary package) and then recompile it as of `cargo run`

**Comment by zzz6519003** [comment]
High-priority

**Comment by weihanglo** [comment]
@zzz6519003 it helps nothing leaving a comment like “High-priority”. At this moment, [cargo-watch](https://crates.io/crates/cargo-watch) and its underlying [watchexec-cli](https://crates.io/crates/watchexec-cli) are good alternatives before a first-party subcommand comes out. They both work seamlessly to me.

---

Copied from [epage's comment](https://internals.rust-lang.org/t/cargo-run-watch-flag-to-auto-recompile/17971/4?u=weihanglo):

> By "adding" cargo watch to cargo, I think integrating it as a `--watch` flag is included in that. Probably first thing is a thorough understanding of `cargo watch` and what works well and doesn't. Besides having a general idea that you want this, we need an understanding of what that means in practice. The discussion should likely continue on that linked issue.

I think it is what to do next if someone wants to give it a try. Further more, you might want to go through a (pre-)RFC process or similar to solicit community feedback. In addition, commands other than `build` can also benefit from *watch mode*, so we should make the scope of the MVP and the future extension clearer. For example, do we want to have Cargo in daemon mode? Is it related to how watch mode is integrated? As a good reference, you can check how [`cargo <file>` `-Zscript`](https://github.com/rust-lang/rfcs/pull/3424) has been done.

---

Triage: in terms of technical details, the MVP needn't be a perfect one. I'd say the efforts level almost reaches “hard” but let's make it a medium one for now. @rustbot: label +E-medium +S-needs-design

**Comment by weihanglo** [comment]
@zzz6519003 it helps nothing leaving a comment like “High-priority”. At this moment, [cargo-watch](https://crates.io/crates/cargo-watch) and its underlying [watchexec-cli](https://crates.io/crates/watchexec-cli) are good alternatives before a first-party subcommand comes out. They both work seamlessly to me.

---

Copied from [epage's comment](https://internals.rust-lang.org/t/cargo-run-watch-flag-to-auto-recompile/17971/4?u=weihanglo):

> By "adding" cargo watch to cargo, I think integrating it as a `--watch` flag is included in that. Probably first thing is a thorough understanding of `cargo watch` and what works well and doesn't. Besides having a general idea that you want this, we need an understanding of what that means in practice. The discussion should likely continue on that linked issue.

I think it is what to do next if someone wants to give it a try. Further more, you might want to go through a (pre-)RFC process or similar to solicit community feedback. In addition, commands other than `build` can also benefit from *watch mode*, so we should make the scope of the MVP and the future extension clearer. For example, do we want to have Cargo in daemon mode? Is it related to how watch mode is integrated? As a good reference, you can check how [`cargo <file>` `-Zscript`](https://github.com/rust-lang/rfcs/pull/3424) has been done.

---

Triage: in terms of technical details, the MVP needn't be a perfect one. I'd say the efforts level almost reaches “hard” but let's make it a medium one for now.
@rustbot label: +E-medium +S-needs-design

**Comment by passcod** [comment]
Hey found out about this thread from someone linking me https://blog.rust-lang.org/inside-rust/2024/02/13/this-development-cycle-in-cargo-1-77.html — happy to answer any queries y'all might have.

**Comment by epage** [comment]
After having used [bacon](https://dystroy.org/bacon/) and seeing feature requests like #11125, my perception has shifted for what I would like to see from this.

I'm imagining cargo commands having a `-i` / `--interactive` flag that launches a full TUI that runs the underlying cargo command, and then re-runs it on changes, with a built-in pager (#11213) that organizes the output and  attempts to keep the view port stable from run-to-run.  Maybe the user could have the `cargo --explain <code>` shown for a given diagnostic, or diagnostics could have a "more details" view that is collapsed or expanded on request.  Maybe the diagnostics themselves could be folded / unfolded between the full view and the short view.  A user could select diagnostics and request for the suggested fix to be applied (#11125).  Maybe there would be an "apply all" option for that category of fix or even all fixes.  For `cargo doc`, the interactive mode could even start a web server (#4966).

It would likely be easiest to prototype something like this outside of cargo, like https://github.com/crate-ci/cargo-fixit and https://github.com/crate-ci/cargo-plumbing, and then evaluate bringing it into cargo.
